core/src/main/scala/akka/persistence/cassandra/query/EventsByTagStage.scala

```diff
@@ -235,24 +252,119 @@ import scala.compat.java8.FutureConverters._
 
       val newResultSetCb = getAsyncCallback[Try[AsyncResultSet]] {
         case Success(rs) =>
-          if (!stageState.state.isInstanceOf[QueryInProgress]) {
-            throw new IllegalStateException(s"New ResultSet when in unexpected state ${stageState.state}")
+          val queryState = stageState.state match {
+            case qip: QueryInProgress => qip
+            case _ =>
+              throw new IllegalStateException(s"New ResultSet when in unexpected state ${stageState.state}")
+          }
+          if (queryState.abortForMissingSearch) {
+            lookForMissing()
+          } else {
+            updateStageState(_.copy(state = QueryResult(rs)))
+            tryPushOne()
           }
-          updateStageState(_.copy(state = QueryResult(rs)))
-          tryPushOne()
         case Failure(e) =>
-          log.warning("Cassandra query failed: {}", e)
+          log.warning("Cassandra query failed: {}", e.getMessage)
           fail(out, e)
       }
 
+      val backTrackCb = getAsyncCallback[Try[Map[PersistenceId, (TagPidSequenceNr, UUID)]]] {
+        case Failure(e) =>
+          updateStageState(_.copy(delayedScanInProgress = false))
+          log.warning("Backtrack failed, this will retried", e)
+        case Success(sequenceNrs) =>
+          updateStageState(_.copy(delayedScanInProgress = false))
+          log.debug("Current sequence nrs: {} from back tracking: {}", stageState.tagPidSequenceNrs, sequenceNrs)
+
+          val withDelayedEvents: Map[PersistenceId, (TagPidSequenceNr, Offset)] = sequenceNrs.filter {
+            case (pid, (tagPidSequenceNr, _)) =>
+              stageState.tagPidSequenceNrs.get(pid) match {
+                case Some((currentTagPidSequenceNr, _, _)) if currentTagPidSequenceNr < tagPidSequenceNr =>
+                  // the current can be bigger than the scanned sequence nrs as the stage continues quiering
+                  // but if the current is lower than the tag pid sequence nr then there is a missed event
+                  true
+                case None =>
+                  // a pid that has never been found before
+                  true
+                case _ =>
+                  false
+              }
+          }
+
+          // TODO: the looking for missing currently tries to find all the events that are missing. What if there is
+          // a large number? Could either fail or change the missing search to find them in chunks and deliver them,
+          // this would mean changing the missing search to find them in order, right now all missing events are found
+          // and then sorted then delivered.
+
+          if (withDelayedEvents.nonEmpty) {
+            log.info(
+              "The following persistence ids have delayed events: {}. Initiating search for the events.",
+              withDelayedEvents)
+
+            stageState match {
```

Comment:
that is one weird pattern match, just do an if-else on isLookingForMissing instead? Also maybe merge it with the surrounding if since it now incorrectly will log that it is initiating a search and then immediate goes "just kidding, I'm not"

Reply:
This started it's life as different branches for all the states then they became one, will switch to if